src/index.ts

```diff
@@ -1,58 +1,255 @@
+    this.setKeypair(kp);
+    return kp;
+  }
+
+  /**
+   * Derive a medusa keypair from a signature
+   * @param {string} signature to use for key derivation
+   * @returns {Keypair<S, P>} A Medusa Keypair
+   */
+  deriveKeypair(signature: string): Keypair<S, P> {
+    // Hashing the signature
+    const hash = utils.keccak256(signature);
+    let random = BigNumber.from(hash);
+    let n = 0;
+    while (true) {
+      try {
+        // Take the scalar from the random number
+        random = random.add(n);
+        const secret = this.suite.scalar().fromEvm(random)._unsafeUnwrap();
+        // Calculate the key: P = G * s
+        const pubkey = this.suite.point().one().mul(secret);
+        return {
+          secret,
+          pubkey,
+        };
+      } catch (e) {
+        n++;
+      }
+    }
+  }
+
+  /**
+   * Fetch the public key of the Medusa Oracle, cache it, and return it
+   * @returns {Promise<P>} The Medusa Public Key
+   */
+  async fetchPublicKey(): Promise<P> {
+    if (this.publicKey) {
+      return this.publicKey;
+    }
+    const medusaContract = ThresholdNetwork__factory.connect(
+      this.medusaAddress,
+      this.signer
+    );
+    const key = await medusaContract.distributedKey();
+    this.publicKey = this.decodePublicKey(key);
+    return this.publicKey;
+  }
+
+  /**
+   * Convert a public key from EVM into Medusa format
+   * @param {EVMG1Point} pubkey to decode
+   * @returns {P} A Medusa Public Key
+   */
+  decodePublicKey(pubkey: EVMG1Point): P {
+    return this.suite.point().fromEvm(pubkey)._unsafeUnwrap();
+  }
+
+  /**
+   * Encrypt a message for a user; include a label to prevent replay attacks via onchain DLEQ verification
+   * @param {Uint8Array} data to encrypt
+   * @param {string} contractAddress of the application developer's contract to be included in the label
+   * @returns {Promise<{encryptedData: string, encryptedKey: HGamalEVMCipher}>} The encrypted data and the EVM encoded encrypted key
+   */
+  async encrypt(
+    data: Uint8Array,
+    contractAddress: string
+  ): Promise<{ encryptedData: string; encryptedKey: HGamalEVMCipher }> {
+    const medusaPublicKey = await this.fetchPublicKey();
+    const hgamalSuite = new HGamalSuite(this.suite);
+    const label = Label.from(
+      medusaPublicKey,
+      contractAddress,
+      await this.signer.getAddress()
+    );
+    const bundle = (
+      await hgamalSuite.encryptToMedusa(data, medusaPublicKey, label)
+    )._unsafeUnwrap();
+
+    return {
+      encryptedKey: bundle.encryptedKey.toEvm(),
+      encryptedData: Base64.fromUint8Array(bundle.encryptedData),
+    };
+  }
+
+  /**
+   * Decrypt a message that has been reencrypted for a user
+   * @param {HGamalEVMCipher} ciphertext of encrypted key to decrypt encryptedContents
+   * @param {string} encryptedContents to decrypt
+   * @returns {Promise<string>} The decrypted data
+   */
+  async decrypt(
+    ciphertext: HGamalEVMCipher,
+    encryptedContents: string
```

Comment:
Same comment as last time no? encrypts takes a uint8array, decrypts should take as well as uint8array (even more knowing that "encryptedContents" is actually encrypted, so it's unreadable bytes) we're not dealing with strings , or should we ? maybe it's a JS thing to prefer dealing with strings and in that case, then it may be even better to only accepts string but that seems very weird...

Reply:
Last time:
encrypt(Uint8array) => string
decrypt(string) => string

Now:
decrypt(string) => Uint8array

I removed an additional step to interpret the bytes as UTF-8.

Why the encrypt/decrypt signatures mismatch is because it's doing base64 encoding/decoding to make it easier for developers. So it's opinionated towards base64 ... which does work well in js ecosystem because of json and json APIs.